icu4c/source/i18n/unitsdata.cpp

```diff
@@ -94,8 +97,213 @@ class ConversionRateDataSink : public ResourceSink {
+                                DecimalQuantity dq;
+                                dq.setToDecNumber(geq.data(), status);
+                                up->geq = dq.toDouble();
+                            } else if (uprv_strcmp(key, "skeleton") == 0) {
+                                int32_t length;
+                                const UChar *s = value.getString(length, status);
+                                up->skeleton.appendInvariantChars(s, length, status);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+  private:
+    MaybeStackVector<UnitPreference> *preferences;
+    MaybeStackVector<UnitPreferenceMetadata> *metadata;
+};
+
+/**
+ * Finds the UnitPreferenceMetadata instance that matches the given category,
+ * usage and region: if missing, region falls back to "001", and usage falls
+ * back to "default".
+ *
+ * This is implemented as a binary search, with fallback restarting the search
+ * from the search range at which the parent in the category/usage/region
+ * hierarchy was found.
+ *
+ * @param metadata The full list of UnitPreferenceMetadata instances.
+ * @param category The category to search for. If category is not known, it can
+ * be resolved from the baseunit of the input (for supported unit categories).
+ * TODO(hugovdm): implement the unit->category lookup (via "unitQuantities" in
+ * the units resource bundle).
+ * @param usage The usage for which formatting preferences is needed. If the
+ * given usage is not known, this function automatically falls back to "default"
+ * usage.
+ * @param region The region for which preferences are needed. If there are no
+ * region-specific preferences, this function automatically falls back to the
+ * "001" region (global).
+ * @param status The standard ICU error code output parameter. If an invalid
+ * category is given, status will be U_ILLEGAL_ARGUMENT_ERROR. If fallback to
+ * "default" or "001" didn't resolve, status will be U_MISSING_RESOURCE.
+ * @return The index into the metadata vector which represents the appropriate
+ * preferences. If appropriate preferences are not found, -1 is returned.
+ */
+int32_t getPreferenceIndex(const MaybeStackVector<UnitPreferenceMetadata> *metadata,
+                           const char *category, const char *usage, const char *region,
+                           UErrorCode &status) {
+    if (U_FAILURE(status)) { return -1; }
+    //  SearchStage stage, const char *category, const char *usage, const char *region) {
+    int32_t start = 0;
+    int32_t end = metadata->length();
+    bool foundCategory = false;
+    bool foundUsage = false;
+    int32_t checkpointStart = start;
+    int32_t checkpointEnd = end;
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        int32_t cmp;
+        cmp = uprv_strcmp((*metadata)[mid]->category.data(), category);
```

Comment:
If you implement `UnitPreferenceMetadata::operator<()` as described above, you can make this binary search code a whole lot shorter and clearer.

Reply:
To not have to redo the binary search more times than necessary, I need to know which components of the "compound primary key" were found and which were not. Since the implementation I snet for review first felt like a sweet spot to me (who had written it and thus can't properly judge), I need some help - which of these two would you prefer:

Simplest code: https://github.com/sffc/icu/pull/42/commits/ab4087b670f3948c4733fc8eaad0b3d4b1b9d279
Same "efficiency" in not performing unnecessary binary searches, but more complexity: https://github.com/sffc/icu/pull/42/commits/d518e3d6e99495baf38743275ea4f274bc2863a4

Reply:
d518e3d is clearer than what you had before, and clear enough for me.

Very optional: another way you could think about this is to get different slices of the array, like this.  Maybe this is over-engineering though.

```cpp
struct UnitPreferenceMetadataSlice {
  UnitPreferenceMetadata** start;
  int32_t length;
  bool isEmpty() const { return length == 0; }

  // Returns a subslice containing all elements with matching categories.
  // If the category does not exist, returns an empty slice.
  Slice forCategory(const char* category) const;

  // Returns a subslice containing all elements with matching usages.
  // If the usage does not exist, returns an empty slice.
  Slice forUsage(const char* usage) const;

  // Returns a subslice containing all elements with matching regions.
  // If the region does not exist, returns an empty slice.
  Slice forRegion(const char* region) const;
}
```